src/main/java/com/google/javascript/clutz/DeclarationGenerator.java

```diff
@@ -1314,16 +1327,48 @@ void walkInnerSymbols(ObjectType type, String innerNamespace) {
         String qualifiedName = innerNamespace + '.' + propName;
         if (provides.contains(qualifiedName)) continue;
         JSType pType = type.getPropertyType(propName);
+
+        Set<JSType> incompatibleTypes = new LinkedHashSet<>();
```

Comment:
I think we should be calling visitProperty here instead of repeating the same code, which your change makes painfully clear. Add a TODO for me if you agree that there is duplicated code.